File: _17_letter_combinations_of_a_phone_number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        # make a dictionary
        # split digits into an array
        # agar digit one ha to os ka map directly return krwa do
        # 2 pointers ka sath nested loop chla do -> on ka .items() sa iterate krwa do

        digit_map = {
            "2": ["a", "b", "c"],
            "3": ["d", "e", "f"],
            "4": ["g", "h", "i"],
            "5": ["j", "k", "l"],
            "6": ["m", "n", "o"],
            "7": ["p", "q", "r", "s"],
            "8": ["t", "u", "v"],
            "9": ["w", "x", "y", "z"],
        }

        digits = list(digits)

        n = len(digits)
        if n == 1:
            return digit_map[digits[0]]

        result = []

        for c1 in digit_map[digits[0]]:
            seq1 = c1
            for c2 in digit_map[digits[1]]:
                seq2 = seq1 + c2
                if n > 2:
                    for c3 in digit_map[digits[2]]:
                        seq3 = seq2 + c3
                        if n > 3:
                            for c4 in digit_map[digits[3]]:
                                seq4 = seq3 + c4
                                result.append(seq4)
                                continue
                        else:
                            result.append(seq3)
                            continue
                else:
                    result.append(seq2)

        return result

File: recursion/_17_letter_combinations_of_a_phone_number.java
public class _17_letter_combinations_of_a_phone_number {
    class Solution {
        Map<Character, List<String>> dict = Map.of(
                '2', List.of("a", "b", "c"),
                '3', List.of("d", "e", "f"),
                '4', List.of("g", "h", "i"),
                '5', List.of("j", "k", "l"),
                '6', List.of("m", "n", "o"),
                '7', List.of("p", "q", "r", "s"),
                '8', List.of("t", "u", "v"),
                '9', List.of("w", "x", "y", "z"));

        public List<String> letterCombinations(String digits) {
            List<String> result = new ArrayList();
            helper("", digits, result);
            return result;
        }

        public void helper(String previous, String digits, List<String> result) {
            if (digits == null || digits.length() == 0) {
                result.add(previous);
                return;
            }

            char digit = digits.charAt(0);
            digits = digits.substring(1);

            for (String ch : dict.get(digit)) {
                helper(previous + ch, digits, result);
            }
        }
    }
}
